src/storage/src/source/reclock.rs

```diff
@@ -8,426 +8,327 @@
-                                .entry(partition)
-                                .or_insert_with(VecDeque::new)
-                                .push_back((
-                                    timestamp,
-                                    MzOffset {
-                                        offset: diff
-                                            .try_into()
-                                            .expect("reclock offset diff to be positive"),
-                                    },
-                                ));
-                        }
-                    }
+                    Ok(Ok(Ok(()))) => break,
+                    Ok(Ok(Err(actual_upper))) => return Err(actual_upper),
+                    Ok(Err(invalid_use)) => panic!("{invalid_use:?}"),
+                    Err(_external_err) => continue,
                 }
             }
+            self.sync(&new_upper).await;
+        }
+        Ok(())
+    }
+
+    /// Calling this method promises that all future timestamping requests will be for offsets
+    /// greater or equal to the provided source compaction frontier. This gives permission to the
+    /// reclock operator to compact and consolidate its internal state accordingly.
+    ///
+    /// If the given source since frontier results in a change of the dest since frontier it will
+    /// be returned to the caller who can then use it to downgrade its capability.
+    pub async fn compact(
+        &mut self,
+        source_since: &HashMap<PartitionId, MzOffset>,
+    ) -> Option<AntichainRef<'_, Timestamp>> {
+        // Translate the SourceTime compaction frontier into a DestTime compaction frontier
+        let new_since = source_since
+            .iter()
+            .map(|(pid, part_since)| {
+                self.partition_uppers(pid)
+                    .find(|(_ts, offset)| part_since < offset)
+                    .expect("attempt to compact past upper")
+                    .0
+            })
+            .min()
+            .map(Antichain::from_elem)
+            .unwrap_or_else(|| self.since.clone());
+
+        if PartialOrder::less_than(&self.since, &new_since) {
+            for bindings in self.remap_trace.values_mut() {
+                // Compact the remap trace according to the computed frontier
+                for (timestamp, _) in bindings.iter_mut() {
+                    timestamp.advance_by(new_since.borrow());
+                }
+                // And then consolidate
+                consolidation::consolidate(bindings);
+            }
+            self.since = new_since;
+            self.read_handle.downgrade_since(self.since.clone()).await;
+            Some(self.since.borrow())
+        } else {
+            None
+        }
+    }
+
+    /// Produces a new timestamp suitable for minting bindings. The returned value is guaranteed to
+    /// be beyond the current `upper` frontier and a multiple of `update_interval_ms`.
+    async fn next_mint_timestamp(&mut self) -> Timestamp {
```

Comment:
happy to push this off until later but we should add a todo and open a task to handle when clock skew between replicas (sorry -- prob wrong terminology?) >> `update_interval_ms`.  

Consider the (prototypical even though possibly not actually applicable here) case of daylight savings time:
- R1, R2, R3 all roughly in sync
- R1 jumps ahead for DST and commits an offset
- R2, R3 read that upper and loop through this function for an hour -- or until their own clocks are updated for DST

I'd imagine that AWS clock skew isn't a _big_ concern here but I think we should make plans to address this in the future to make our timestamping more robust

Reply:
DST don't translate to changes to the unix timestamps but your point is definitely valid for clock skew. What would be the defense though? Like, how do we know that our clock is bad? And what else could we do other than wait an hour if indeed someone commits an offset an hour in the future?

Reply:
I think the fix would be improving our clock, i.e., what `(self.now)()` returns.  Or perhaps if the jump is sufficiently large, just panic the current replica and then spin a new one up -- that could arguably be a better user experience than hanging for a long time.

Just a TODO so we revisit later -- I don't think we need (or want!) to try and address it now